File: Readio/src/main/java/ninegle/Readio/subscription/domain/Subscription.java
package ninegle.Readio.subscription.domain;

import java.time.LocalDate;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import lombok.AccessLevel;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;

@Entity
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
public class Subscription {

	@Id
	@GeneratedValue(strategy = GenerationType.IDENTITY)
	private Long id;

	@Column(unique = true) // 유저당 하나의 구독만 허용
	private Long userId;

	private LocalDate subDate;

	private LocalDate expDate;

	@Column(nullable = false)
	private boolean canceled;

	@Builder
	public Subscription(Long userId, LocalDate subDate, LocalDate expDate, boolean canceled) {
		this.userId = userId;
		this.subDate = subDate;
		this.expDate = expDate;
		this.canceled = false; // 기본값 설정
	}

	// 구독 기간 갱신
	public void updatePeriod(LocalDate subDate, LocalDate expDate) {
		this.subDate = subDate;
		this.expDate = expDate;
	}

	// 구독 취소
	public void cancel() {
		this.canceled = true;
	}

	// 구독 취소 상태 해제
	public void uncancel() {
		this.canceled = false;
	}

	// 구독이 유효한지 확인
	public boolean isActive() {
		return expDate.isAfter(LocalDate.now()) || expDate.isEqual(LocalDate.now());
	}
}

File: Readio/src/main/java/ninegle/Readio/subscription/domain/Subscription.kt
package ninegle.Readio.subscription.domain

import jakarta.persistence.*
import java.time.LocalDate

@Entity
open class Subscription(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    val id: Long? = null,

    @Column(unique = true)
    val userId: Long,

    var subDate: LocalDate,

    var expDate: LocalDate,

    @Column(nullable = false)
    var canceled: Boolean = false
) {
    //기본생성자
    constructor() : this(
        id = null,
        userId = 0L,
        subDate = LocalDate.now(),
        expDate = LocalDate.now(),
        canceled = false
    )

   fun updatePeriod(subDate: LocalDate, expDate: LocalDate) {
        this.subDate = subDate
        this.expDate = expDate
    }

    fun cancel() {
        this.canceled = true
    }

    fun uncancel() {
        this.canceled = false
    }

    fun isActive(): Boolean =
        expDate.isAfter(LocalDate.now()) || expDate.isEqual(LocalDate.now())

    // equals
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is Subscription) return false
        return id != null && id == other.id
    }

    //hashCode
    override fun hashCode(): Int = id?.hashCode() ?: 0

    override fun toString(): String {
        return "Subscription(id=$id, userId=$userId, subDate=$subDate, expDate=$expDate, canceled=$canceled)"
    }

    //새 구독 생성
    companion object {
        fun create(userId: Long, subDate: LocalDate, expDate: LocalDate): Subscription {
            return Subscription(userId = userId, subDate = subDate, expDate = expDate)
        }
    }


}